commit 82fe66348833aad00379a39e9a116f927fb33dfe
Author:     nvkevlu <55759229+nvkevlu@users.noreply.github.com>
AuthorDate: Thu Nov 21 08:59:33 2024 -0800
Commit:     GitHub <noreply@github.com>
CommitDate: Thu Nov 21 11:59:33 2024 -0500

    Add ability for clients to send error log to server (#3057)
    
    * have client send error log to server
    
    * rename file and fix ci
    
    * fix ci
    
    * revise PR
    
    * fix formatting
    
    ---------
    
    Co-authored-by: Yuan-Ting Hsieh (謝沅廷) <yuantingh@nvidia.com>
---
 nvflare/private/fed/client/client_executor.py | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/nvflare/private/fed/client/client_executor.py b/nvflare/private/fed/client/client_executor.py
index 898dd2e9..ac67006d 100644
--- a/nvflare/private/fed/client/client_executor.py
+++ b/nvflare/private/fed/client/client_executor.py
@@ -179,7 +179,7 @@ class JobExecutor(ClientExecutor):
 
         thread = threading.Thread(
             target=self._wait_child_process_finish,
-            args=(client, job_id, allocated_resource, token, resource_manager, args.workspace),
+            args=(client, job_id, allocated_resource, token, resource_manager, args.workspace, fl_ctx),
         )
         thread.start()
 
@@ -384,7 +384,9 @@ class JobExecutor(ClientExecutor):
             )
             self.logger.debug("abort_task sent")
 
-    def _wait_child_process_finish(self, client, job_id, allocated_resource, token, resource_manager, workspace):
+    def _wait_child_process_finish(
+        self, client, job_id, allocated_resource, token, resource_manager, workspace, fl_ctx
+    ):
         self.logger.info(f"run ({job_id}): waiting for child worker process to finish.")
         job_handle = self.run_processes.get(job_id, {}).get(RunProcessKey.JOB_HANDLE)
         if job_handle:
@@ -393,6 +395,7 @@ class JobExecutor(ClientExecutor):
             return_code = get_return_code(job_handle, job_id, workspace, self.logger)
 
             self.logger.info(f"run ({job_id}): child worker process finished with RC {return_code}")
+
             if return_code in [ProcessExitCode.UNSAFE_COMPONENT, ProcessExitCode.CONFIG_ERROR]:
                 request = new_cell_message(
                     headers={},
@@ -419,6 +422,11 @@ class JobExecutor(ClientExecutor):
             self.run_processes.pop(job_id, None)
         self.logger.debug(f"run ({job_id}): child worker resources freed.")
 
+        engine = fl_ctx.get_engine()
+        fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, job_id, private=True, sticky=False)
+        fl_ctx.set_prop(FLContextKey.CLIENT_NAME, client.client_name, private=True, sticky=False)
+        engine.fire_event(EventType.JOB_COMPLETED, fl_ctx)
+
     def get_status(self, job_id):
         process_status = self.run_processes.get(job_id, {}).get(RunProcessKey.STATUS, ClientStatus.STOPPED)
         return process_status
